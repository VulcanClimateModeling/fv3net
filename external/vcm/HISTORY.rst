=======
History
=======

Current
-------

New Features:

- new public function `vcm.open_remote_nc`
- new public class `vcm.ZarrMapping` for a database-like front-end to zarr
- Added dask support for VCM's vertical remapping function,
  :py:func:`vcm.cubedsphere.regridz.regrid_vertical`.

0.1.0 (2020-03-20)
------------------
First VCM release (w/ fv3net for now)
