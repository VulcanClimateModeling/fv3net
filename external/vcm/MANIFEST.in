include pyproject.toml
include HISTORY.rst
include LICENSE
include README.rst
include vcm/catalog.yaml

recursive-include tests *
recursive-exclude * __pycache__
recursive-exclude * *.py[co]

recursive-include docs *.rst conf.py Makefile make.bat *.jpg *.png *.gif
