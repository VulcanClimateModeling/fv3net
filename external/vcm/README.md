VCM Tools
=========

VCM Tools is a collection of routines used by the Vulcan Climate Modeling group.

* Free software: BSD license

For more information see the [documentation](https://vulcanclimatemodeling.com/docs/vcm).
