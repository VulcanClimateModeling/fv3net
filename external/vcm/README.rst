FV3Config
=========


FV3Config is used to configure and manipulate run directories for FV3GFS.

* Free software: BSD license

Basic usage
-----------

.. code-block:: python

    from fv3config import get_default_config, write_run_directory

    config = get_default_config()
    write_run_directory(config, './rundir')

:code:`config` is a configuration dictionary which contains namelists, input data specifications,
and other options. It can be edited just like any dictionary.

For more in-depth usage, please refer to the documentation. This can be generated with :code:`make docs`.
