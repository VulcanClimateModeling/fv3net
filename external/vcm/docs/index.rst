VCM Tools Documentation
=======================

VCM Tools is a collection of routines used by the AI2 Climate Modeling group.
It is licensed under a BSD License.

.. toctree::
   :maxdepth: 1
   :caption: Setup

   installation

.. toctree::
   :maxdepth: 1
   :caption: Tools

   calculations
   cloud
   fv3
   combining
   rotating-winds
   sampling
   derived-mapping
   grid-operations
   interpolation
   xarray-io
   xarray-tools
   thermodynamics
   times

.. toctree::
   :maxdepth: 1
   :caption: Reference

   history

Indices and tables
==================
* :ref:`genindex`
* :ref:`modindex`
* :ref:`search`
