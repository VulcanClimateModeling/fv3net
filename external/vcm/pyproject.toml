# credit: https://stackoverflow.com/questions/60045913/installing-numpy-before-using-numpy-distutils-core-setup
[build-system]
build-backend = "setuptools.build_meta"
requires=["numpy==1.19.4", "setuptools==49.6.0"]