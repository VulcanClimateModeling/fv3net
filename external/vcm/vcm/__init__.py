from .combining import combine_array_sequence

from . import cubedsphere
from .extract import extract_tarball_to_path
from .fv3_restarts import open_restarts
from .convenience import (
    TOP_LEVEL_DIR,
    parse_timestep_str_from_path,
    parse_datetime_from_str,
)
from .coarsen import coarsen_restarts_on_pressure, coarsen_restarts_on_sigma
from .visualize import plot_cube, mappable_var, plot_cube_axes

from .xarray_loaders import open_tiles

__all__ = [
    "TOP_LEVEL_DIR",
    "calc",
    "casting",
    "cloud",
    "coarsen",
    "coarsen_restarts_on_pressure",
    "coarsen_restarts_on_sigma",
    "combine_array_sequence",
    "combining",
    "complex_sfc_data_coarsening",
    "convenience",
    "cubedsphere",
    "extract",
    "extract_tarball_to_path",
    "fv3_restarts",
    "mappable_var",
    "open_restarts",
    "open_tiles",
    "parse_datetime_from_str",
    "parse_timestep_str_from_path",
    "plot_cube",
    "plot_cube_axes",
    "regrid",
    "schema",
    "schema_registry",
    "visualize",
    "xarray_loaders",
    "xarray_utils",
]
