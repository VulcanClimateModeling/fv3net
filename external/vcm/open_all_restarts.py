from vcm.convenience import open_delayed
import vcm
from dask.delayed import delayed
import pandas as pd
import xarray as xr
import fsspec
import os
from pathlib import Path

from vcm.fv3_restarts import _parse_time_string

# from distributed import Client

# client = Client("tcp://192.168.100.88:8786")


@delayed
def open_restarts_eager(url):
    return vcm.open_restarts_with_time_coordinates(url).load()


def change_to_forecast_time(ds):
    ds_dt = ds.assign_coords(time=ds.time - ds.time[0])
    return ds_dt.rename({'time': 'forecast_time'})


def _parse_time_from_base(path):
    timestep = str(Path(path).name)
    return _parse_time_string(timestep)




fs = fsspec.filesystem('gs')

url = "gs://vcm-ml-data/orchestration-testing/test-andrep/one_step_run_one_step_yaml_all-physics-off.yml_experiment_label_test-orchestration-group/"


if __name__ == '__main__':
    timesteps = ['gs://' + _ for _ in fs.ls(url)]

    initial_times = []
    delayed_objs = []
    run_folder = []

    for step in timesteps:
        try:
            time = _parse_time_from_base(step)
        except ValueError:
            pass
        else:
            obj = open_restarts_eager(step)
            delayed_objs.append(obj)
            initial_times.append(time)
            run_folder.append(step)


    # load one file shema
    schema = vcm.open_restarts_with_time_coordinates(run_folder[0])

    # apply metadata
    datasets = [open_delayed(obj, schema) for obj in delayed_objs]

    # change time to delta time
    datasets_no_time = [change_to_forecast_time(ds) for ds in datasets]

    merged = xr.concat(
        datasets_no_time,
        dim=pd.Index(run_folder, name='run_folder')
    )

    initial_times = xr.DataArray(initial_times, dims=['run_folder'])

    cleaned = merged.assign_coords(initial_time=initial_times).swap_dims({'run_folder': 'initial_time'})

    # cleaned.to_zarr("big.zarr", mode="w")
