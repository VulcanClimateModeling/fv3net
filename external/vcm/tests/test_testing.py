import warnings
import pytest

from vcm import testing


def test_no_warning():
    with pytest.raises(AssertionError):
        with testing.no_warning(None):
            warnings.warn("Warning")
