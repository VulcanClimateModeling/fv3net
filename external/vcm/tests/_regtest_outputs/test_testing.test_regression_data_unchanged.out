Array hash:
9ec20a9830adbe3d9a424eeea58ebecc
Coordinate info:
Coordinate  x : [1]
xarray.Dataset {
dimensions:
	x = 1 ;

variables:
	int64 x(x) ;
	int64 a(x) ;
		a:hello = world ;

// global attributes:
}
